File: backend/tiramisu-lite/Model/ShoppingList.cs
namespace tiramisu_lite.Model;

public class ShoppingList
{
    public Guid Id { get; private set; }
    public Guid ProfileId { get; private set; }
    public Profile Profile { get; init; } = null!;
    public string Name { get; private set; }
    public DateTime CreatedAt { get; private set; }

    public ShoppingList(Guid id, Guid profileId, string name, DateTime createdAt)
    {
        ArgumentNullException.ThrowIfNull(createdAt, nameof(createdAt));
        this.Id = id;
        this.ProfileId = profileId;
        this.Name = name;
        this.CreatedAt = createdAt;
    }

    public required IEnumerable<ShoppingListItem> ShoppingListItems
    {
        get => this.shoppingListItems;
        init => this.shoppingListItems = value.ToList();
    }

    private List<ShoppingListItem> shoppingListItems;

    public bool Completed => this.ShoppingListItems.All(item => item.Completed);

    public void UpdateName(string name)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(name, nameof(this.Name));
        this.Name = name;
    }
}

File: mobile/lib/model/shopping_list.dart
import 'package:mobile/model/shopping_list_item.dart';

class ShoppingList {
  final String id;
  final String name;
  final List<ShoppingListItem> items;
  final DateTime createdAt;

  ShoppingList({
    required this.id,
    required this.name,
    required this.items,
    required this.createdAt,
  });

  factory ShoppingList.fromJson(Map<String, dynamic> json) {
    return ShoppingList(
      id: json['id'],
      name: json['name'],
      createdAt: DateTime.parse(json['createdAt']),
      items:
          (json['items'] as List)
              .map((item) => ShoppingListItem.fromJson(item))
              .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'createdAt': createdAt.toIso8601String(),
      'items': items.map((item) => item.toJson()).toList(),
    };
  }

  ShoppingList copyWith({
    String? id,
    String? name,
    List<ShoppingListItem>? items,
    DateTime? createdAt,
  }) {
    return ShoppingList(
      id: id ?? this.id,
      name: name ?? this.name,
      items: items ?? this.items,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}
